examples/src/fungible-token.js

```diff
@@ -1,87 +1,136 @@
-import {
-  NearBindgen,
-  call,
-  view,
-  initialize,
-  near,
-  LookupMap,
-  assert,
-} from "near-sdk-js";
+import { NearBindgen, call, view, initialize, near, LookupMap, assert } from "near-sdk-js";
 
 @NearBindgen({ initRequired: true })
 export class FungibleToken {
-  constructor() {
-    this.accounts = new LookupMap("a");
-    this.totalSupply = 0n;
-  }
+  accounts = new LookupMap("a");
+  accountRegistrants = new LookupMap("r");
+  accountDeposits = new LookupMap("c");
+  totalSupply = "0";
 
   @initialize({})
-  init({ prefix, totalSupply }) {
-    this.accounts = new LookupMap(prefix);
-    this.totalSupply = BigInt(totalSupply);
-    this.accounts.set(near.signerAccountId(), this.totalSupply);
-    // In a real world Fungible Token contract, storage management is required to denfense drain-storage attack
+  init({ owner_id, total_supply }) {
+    assert(BigInt(total_supply) > BigInt(0), "Total supply should be a positive number");
+    assert(this.totalSupply === "0", "Contract is already initialized");
+    this.totalSupply = total_supply;
+    this.accounts.set(owner_id, this.totalSupply);
+  }
+
+  internalGetMaxAccountStorageUsage() {
+    const initialStorageUsage = near.storageUsage();
+    const tempAccountId = "a".repeat(64);
+    this.accounts.set(tempAccountId, "0");
+    const maxAccountStorageUsage = near.storageUsage() - initialStorageUsage;
+    this.accounts.remove(tempAccountId);
+    return maxAccountStorageUsage * BigInt(3); // we create an entry in 3 maps
   }
 
-  internalDeposit({ accountId, amount }) {
-    let balance = this.accounts.get(accountId, { defaultValue: 0n });
-    let newBalance = balance + BigInt(amount);
-    this.accounts.set(accountId, newBalance);
-    this.totalSupply += BigInt(amount);
+  internalRegisterAccount({ registrantAccountId, accountId, amountStr }) {
+    assert(!this.accounts.containsKey(accountId), "Account is already registered");
+    this.accounts.set(accountId, "0");
+    this.accountRegistrants.set(accountId, registrantAccountId);
+    this.accountDeposits.set(accountId, amountStr);
   }
 
-  internalWithdraw({ accountId, amount }) {
-    let balance = this.accounts.get(accountId, { defaultValue: 0n });
-    let newBalance = balance - BigInt(amount);
-    assert(newBalance >= 0n, "The account doesn't have enough balance");
-    this.accounts.set(accountId, newBalance);
-    let newSupply = this.totalSupply - BigInt(amount);
-    assert(newSupply >= 0n, "Total supply overflow");
-    this.totalSupply = newSupply;
+  internalSendNEAR(receivingAccountId, amountBigInt) {
+    assert(amountBigInt > BigInt("0"), "The amount should be a positive number");
+    assert(receivingAccountId != near.currentAccountId(), "Can't transfer to the contract itself");
+    assert(amountBigInt < near.accountBalance(), `Not enough balance ${near.accountBalance()} to cover transfer of ${amountBigInt} yoctoNEAR`);
+    const transferPromiseId = near.promiseBatchCreate(receivingAccountId);
```

Comment:
Try to use high-level promise API here. We have examples in the repo.

Reply:
Sounds good: implemented